### ISSUE TITLE
Doublons non reconnus

### ISSUE THREAD

**Issue by EtienneM** [bnjbvr/kresus#362]
Salut,

Petit problème avec la reconnaissance des doublons sur mon Kresus. C'est la deuxième fois en peu de temps que j'ai pas mal de doublons mais aucun n'est reconnu dans la section "Doublon". Du coup, la balance en cours de mon compte n'est plus correct... Une idée ? 

Pour info, mon kresus est à jour. Je suis hébergé chez CozyCloud.

**Comment by ZeHiro** [comment]
Bonjour @EtienneM ,
Kresus détecte les doublons de cette manière : 
- si la valeur des opérations est la même et
- si la différence entre les dates des l'opérations est inférieure à la valeur seuil définie dans les paramètres et
- si les opérations ne sont pas importées à la même date

Il semblerait que le problème vienne de la dernière condition qui n'est pas remplie. 
Donc, weboob doit fournire des doublons, mais il me semble que si weboob fournit des opérations "doublons" elle ne sont pas importées. @bnjbvr tu confirmes? Ce ne serait pas un nouvel effet de bord de la fusion automatique des comptes?

**Comment by bnjbvr** [comment]
Salut, ça pourrait aussi être la différence de dates par défaut qui est trop faible. As tu essayé d'augmenter le seuil de détection de doublons dans les préférences ? (ce n'est pas une option simple à trouver, je commence à me dire qu'elle devrait être dans les doublons directement, ou cachée derrière un bouton "trouver plus de doublons").

**Comment by EtienneM** [comment]
J'importe systématiquement mes opérations manuellement. Je ne sais pas pourquoi mais l'importation ne se fait jamais automatiquement avec mon Kresus hébergé chez Cozy Cloud.

Je viens de faire un tour dans le data browser. J'ai par exemple une opération qui apparait 2 fois. La date d'import est pour les deux lignes identique : 2016-01-25T09:06:01.000Z. Le seuil de doublon ne change rien

**Comment by EtienneM** [comment]
up

**Comment by bnjbvr** [comment]
C'est très étrange. Deux opérations en double importées au même moment ne devraient pas être des doublons ; est-ce que par chance, il s'agirait bien de deux opérations distinctes ?

Sinon, ça pourrait être effectivement une conséquence malheureuse de la fusion de comptes, mais même dans ce cas, je ne vois pas pourquoi kresus aurait importé des opérations appartenant à deux comptes différents puis auraient décidé de les fusionner après. A un moment donné, Kresus n'aurait vu que les opérations de l'ancien compte ou du nouveau compte, mais pas les deux en même temps. 

A moins qu'il ne s'agisse de deux comptes différents ayant le même nom et même identifiant interne, et que ces deux comptes aient eus tous les deux la même opération. Est-ce ton cas, Etienne ?

**Comment by ZeHiro** [comment]
@bnjbvr, un probleme du module weboob?

**Comment by bnjbvr** [comment]
Oui, c'est une autre possibilité, ça serait vraiment pas de chance par contre...

**Comment by EtienneM** [comment]
J'ai fait des tests en local avec [boobank](http://weboob.org/applications/boobank), l'application banque de Weboob. Aucun doublon n'y apparait donc ça serait plutôt un problème avec Kresus. 

> est-ce que par chance, il s'agirait bien de deux opérations distinctes ?

Qu'entends-tu par là ?

Je n'ai jamais fait de fusion de compte, ce compte est configuré depuis que j'utilise Kresus et mon numéro de compte n'a jamais changé. Est-ce qu'il y a des logs quelque part que je pourrais t'envoyer pour t'aider ? Est-ce que tu as accès à la plateforme d'hébergement de Cozy vu que je suis hébergé chez eux ?

**Comment by bnjbvr** [comment]
Par "deux opérations distinctes", je demandais s'il était possible que ce soit bien deux opérations différentes dans ta banque, mais avec les mêmes libellés, dates et montants.

Je n'ai pas accès aux logs de Cozy (je ne travaille pas pour la boîte). Les logs serveurs risquent de ne pas apporter beaucoup d'information, surtout si tu as redémarré kresus récemment (les logs sont surchargés à chaque redémarrage). Si jamais tu as une grosse quantité de logs, me les envoyer par email pourrait être sympa (cf l'adresse email que j'utilise pour committer).

Quoi qu'il en soit, j'ai l'impression qu'il s'agit d'une erreur assez rare, pas gérée. Je prendrai soin d'auditer tout le code qui pull les opérations, mais je n'ai vraiment aucune idée sur ce qui aurait pu arriver (à part une fusion bizarre de deux comptes). Pour contourner le problème, je t'invite à supprimer les opérations en doublons à la main dans le databrowser, mais ça risque d'être fastidieux... Sinon, je peux te faire une toute petite version de kresus rien que pour toi, temporairement, sur une branche séparée, qui ne prendrait pas en compte la date d'import dans la détection des doublons, juste pour que tu puisses fusionner tes doublons facilement.

Une autre solution consiste à exporter ton instance ; repérer les doublons à la main dans le fichier exporté, les supprimer ; supprimer tous tes comptes ; réimporter ton instance. Mais attention, c'est dangereux, et pas couvert par la garantie ;)

**Comment by EtienneM** [comment]
OK merci pour toutes ces infos. Pour le moment, rien d'intéressants dans les logs. Dès que j'ai cette erreur à nouveau, je regarderai les logs. En attendant je vais supprimer dans le data browser les opérations en double ;)

**Comment by EtienneM** [comment]
J'ai à nouveau eu le souci en mettant à jour mes comptes ce matin. J'ai mis les logs ici : https://bimestriel.framapad.org/p/r.4c92536dbd53a1283db650cf069d49c7

Pour info, j'avais configuré un seuil d'alerte sur solde pour un de mes comptes. J'ai reçu l'email d'alerte 3 fois. Peut-être est-ce lié.

**Comment by bnjbvr** [comment]
Merci pour l'info ! Si jamais tu as encore ces opérations dans un coin, est-ce que tu pourrais recopier leur contenu entier (depuis le databrowser) et le transmettre, stp ? (tu peux changer les textes et montants, tant que la structure est la même, c'est à dire même texte ou pas, même montant ou pas, etc.)

**Comment by EtienneM** [comment]
OK je t'envoie ça par email.

**Comment by ZeHiro** [comment]
@bnjbvr , y a t-il des nouvelles sur ce point? Il a à nouveau été observé au crédit coopératif

**Comment by bnjbvr** [comment]
Pas vraiment de nouvelles, je ne sais pas trop ce qu'il se passe. En relisant les logs framapad ci dessus, j'ai l'impression qu'il y a eu deux synchros au démarrage de Kresus. Si les deux synchros arrivent en même temps, il y a effectivement une course (race condition) et toutes les opérations récupérées lors de ces synchros pourraient être prises en double. Je vais rajouter une petite protection pour mitiger cela dans la prochaine version.
